### PR TITLE
Features/fastapi integration

### PR DESCRIPTION
**Pull request by dwrtz**
### Background

This change allows Auto-GPT to be run from a server. It supports multiple users running agents simultaneously. See `Server.md` for instructions.

It has a very basic HTTP API with 2 POST endpoints (/start, and /chat).

This is a prerequisite to building a browser/mobile front-end.

### Changes

The server:
- app.py, app_types.py, and app_utils.py
- these implement the server. It's very similar to the interaction loop in main.py. However, instead of getting user input from stdin, user input comes from http post requests.

Supporting multiple users:
- I introduced the concept of a chat_id. This lets the server keep track of which conversation belongs to which user.
- This has implications for pinecone memory. Now all the memory associated with particular chat_id lives in its own namespace.
- This also has implications for the filesystem. The files relevant to a particular chat will be created in ./auto_gpt_workspace/<chat_id>
- Lots of files had to be modified to support this. The changes were mostly very simple, and involved plumbing the chat_id through functions to where it needed to be. See commands.py for an idea of what was needed.

Previous functionality is maintained! You can still run main.py and it will work exactly like it used to.

### Test Plan

I used postman to test the api locally. I also ran main.py to make sure that it still works as expected. 

That said, I haven't tested a huge variety of prompts. Mostly only prompts that write files.

Here's my exported postman collection:
```json
{
	"info": {
		"_postman_id": "68494d7e-b31d-4de1-8e2d-48b4d13cb465",
		"name": "auto-gpt",
		"schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json",
		"_exporter_id": "19337810"
	},
	"item": [
		{
			"name": "start",
			"request": {
				"method": "POST",
				"header": [],
				"body": {
					"mode": "raw",
					"raw": "{\n    \"ai_name\": \"HelloBot\",\n    \"ai_role\": \"An AI that says 'Hello, World!'\",\n    \"ai_goals\": [\n        \"Write your message in a file called 'message.txt'.\",\n        \"Shut down.\"\n    ]\n}",
					"options": {
						"raw": {
							"language": "json"
						}
					}
				},
				"url": {
					"raw": "localhost:8000/start",
					"host": [
						"localhost"
					],
					"port": "8000",
					"path": [
						"start"
					]
				}
			},
			"response": []
		},
		{
			"name": "chat",
			"request": {
				"method": "POST",
				"header": [
					{
						"key": "chat_id",
						"value": "33178f06-2c4c-4db4-96b6-f1e1cd6bd6b3",
						"type": "text"
					}
				],
				"body": {
					"mode": "raw",
					"raw": "{\n    \"message\": \"y\"\n}",
					"options": {
						"raw": {
							"language": "json"
						}
					}
				},
				"url": {
					"raw": "localhost:8000/chat",
					"host": [
						"localhost"
					],
					"port": "8000",
					"path": [
						"chat"
					]
				}
			},
			"response": []
		}
	]
}
```

### Change Safety

- [x] I have added tests to cover my changes
  - Ideally we'd have a series of tests for each command in commands.py
- [x] I have considered potential risks and mitigations for my changes
  - This is not ready for production. For at least a few reasons...
    - No user authentication
    - File IO writes to disk. Ideally it should probably mount a network volume, possibly per chat_id.
    - Uses python dictionaries as key/value storage for chat context in between requests. Eventually should use a real database.

### CODE REVIEW

**Review by JukoiDev — APPROVED** [review]
LGTM, very clean code

**Comment by merwanehamadi** [comment]
Thanks a lot ! I left a comment.
It looks like for now there is no loop right ? It's like a manual mode if I understand ? it's actually very good to have it to get very granular control over it, and then maybe during the chat you tell him "alright do your thing for a minute or two and come back to me.

This also eliminates my concern: I was worried because we didn't have a 
PATCH /autonomous_agent/{id}
{
  "status": "inactive"
}
to kill the agent.

But here it's not needed.

**Review by JustMrMendez — APPROVED** [review]

**Comment by merwanehamadi** [comment]
```
@app.post("/start")
```

This is not the right endpoint for a REST API, you shouldn't use verbs. And when you post the /agent, do not start the chat with it, you should just save the AIConfig and then have another endpoint to start the chat. Also, this whole thing only works with manual agents, because you're not starting the loop. Overall I don't think we should merge this, this is very targeted for one use case and not very general.
See what I did here: 
https://github.com/Torantulino/Auto-GPT/pull/765/files

Please wait for my PR to get merged and reuse the POST /agents endpoint. Then you can add your post /chat endpoint. and I would rename it /conversation like openAI's endpoint, this will simplify frontend integration because with "chat" we don't know if it's a verb or a name.
Before my PR gets merged I will create a PR for proper modules like what @ryanpeach tried to do in vain.

**Pull request closed without merging**
